aws-janitor/resources/s3_bucket.go

```
@@ -0,0 +1,141 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type S3Bucket struct{}
+
+// IsManagedS3Bucket checks whether the bucket should be managed by janitor.
+func IsManagedS3Bucket(opts Options, region string, bucketName string) (bool, error) {
+	svc := s3.New(opts.Session, &aws.Config{Region: aws.String(region)})
+	respTags, err := svc.GetBucketTagging(&s3.GetBucketTaggingInput{Bucket: aws.String(bucketName)})
+	if err != nil {
+		return false, err
+	}
+	tags := fromS3Tags(respTags.TagSet)
+	return opts.ManagedPerTags(tags), nil
+}
+
+func (S3Bucket) MarkAndSweep(opts Options, set *Set) error {
+	logger := logrus.WithField("options", opts)
+	if !opts.EnableS3BucketClean {
+		logger.Info("Disable s3 bucket clean")
+		return nil
+	}
+	svc := s3.New(opts.Session, &aws.Config{Region: aws.String(opts.Region)})
+
+	// ListBucket lists all buckets owned by the authenticated sender of the request
+	// https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#S3.ListBuckets
+	// regardless the region configured in client.
+	resp, err := svc.ListBuckets(&s3.ListBucketsInput{})
+	if err != nil {
+		return err
+	}
+	for _, b := range resp.Buckets {
+		bucketName := *b.Name
+
+		region, err := s3manager.GetBucketRegion(context.Background(), opts.Session, bucketName, opts.Region)
+		if err != nil || region != opts.Region {
+			continue
```

Can we/should we log errors?  Otherwise it's hard to diagnose what's going wrong if buckets are not deleted...

---

I don't think we should log errors here, the reason is that "ListBuckets" method lists buckets of all regions, and GetBucketRegion can only get regions near "hint region"
https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#GetBucketRegion
In this case, if we log error here, most of our log will be filled with the "region error".
To make the error log easier to read, I prefer to let this part fail silently.

---

OK, that makes sense.  A comment as to why might help others avoid adding this in future :-)